// hifi-spatial-audio/dist/index.js
import { HiFiUtilities } from "./utilities/HiFiUtilities";
export { HiFiUtilities };
let isBrowserContext = typeof self !== 'undefined';
if (isBrowserContext) {
    HiFiUtilities.checkBrowserCompatibility();
}
let apiVersion = typeof (HIFI_API_VERSION) === "string" ? HIFI_API_VERSION : "unknown";
export { apiVersion as hiFiAPIVersion };
export { apiVersion };
export { HiFiAudioAPIData, ReceivedHiFiAudioAPIData, OrientationEuler3D, OrientationQuat3D, Point3D, eulerToQuaternion, eulerFromQuaternion } from "./classes/HiFiAudioAPIData";
export { HiFiCommunicator, HiFiConnectionStates, HiFiUserDataStreamingScopes } from "./classes/HiFiCommunicator";
export { AvailableUserDataSubscriptionComponents, UserDataSubscription } from "./classes/HiFiUserDataSubscription";
export { HiFiLogLevel, HiFiLogger } from "./utilities/HiFiLogger";
export { HiFiConstants } from "./constants/HiFiConstants";
export { HiFiAxes, HiFiHandedness, HiFiAxisConfiguration } from "./classes/HiFiAxisConfiguration";
let getBestAudioConstraints = HiFiUtilities.getBestAudioConstraints;
export { getBestAudioConstraints };
let preciseInterval = HiFiUtilities.preciseInterval;
export { preciseInterval };
export { HiFiCommunicator as Communicator, HiFiConnectionStates as ConnectionStates, HiFiUserDataStreamingScopes as UserDataStreamingScopes } from "./classes/HiFiCommunicator";
export { ReceivedHiFiAudioAPIData as ReceivedAudioAPIData, HiFiAudioAPIData as AudioAPIData } from "./classes/HiFiAudioAPIData";
export { HiFiLogger as Logger, HiFiLogLevel as LogLevel } from "./utilities/HiFiLogger";
export { HiFiUtilities as Utilities } from "./utilities/HiFiUtilities";
export { HiFiConstants as Constants } from "./constants/HiFiConstants";
export { HiFiAxes as Axes, HiFiHandedness as Handedness, HiFiAxisConfiguration as AxisConfiguration } from "./classes/HiFiAxisConfiguration";
